test/consumer/message_tracker_test.exs

```diff
@@ -197,29 +185,21 @@ defmodule Mississippi.Consumer.MessageTracker.Test do
 
       MessageTracker.handle_message(message_tracker_pid, message_1, channel)
       MessageTracker.handle_message(message_tracker_pid, message_2, channel)
-      message_queue = :queue.from_list([message_1, message_2])
 
-      assert %State{queue: ^message_queue} = :sys.get_state(message_tracker_pid)
-
-      assert_receive {:trace, ^data_updater_pid, :receive, {_, {:handle_message, ^message_1}}}
-
-      refute_received {:trace, ^data_updater_pid, :receive, {_, {:handle_message, ^message_2}}}
```

Comment:
Just as above, the `refute_received` makes sure that the data updater has not received the second message before it finished handling the first. I would keep that

Reply:
refute_receive does not assert that no message was received before the one you're testing, only that in the timeout (100ms) it did not receive that specific message. As above, it happened that both messages were received in the timeout. In my opinion, we should only make assertions on the ordering and not the timings

Reply:
The issue here is that `assert_receive` does not make an assertion about ordering, as `receive` blocks go through the whole process mailbox trying to find a match (see https://hexdocs.pm/elixir/processes.html#sending-and-receiving-messages). `refute_received` provided a guarantee that a given message was not received, thus we could infer an ordering.
I'm open to finding another solution, if that was the reason for flakiness